src/number_theory/bernoulli.lean

```diff
@@ -257,3 +257,120 @@ begin
+    { by_contradiction g,
+      simp only [not_not, nat.cast_succ, nat.factorial_succ, nat.cast_eq_zero,
+        nat.cast_mul, mul_eq_zero] at g,
+      cases g,
+      { exact nat.cast_add_one_ne_zero q g,},
+      { exact nat.factorial_ne_zero q g, }, },
+    rw [eq_comm],
+    rw [div_eq_iff hq],
+    have hqm1: q - m + 1 = q + 1 - m,
+    { rw [nat.sub_add_comm],
+      apply nat.le_of_lt_succ,
+      exact h, },
+    rw [nat.succ_eq_add_one, mul_assoc _ _ ↑(q.succ.factorial),
+      mul_comm  _ ↑(q.succ.factorial), ←mul_assoc, div_mul_eq_mul_div,
+      mul_comm (n ^ (q -m + 1):ℚ) _, ←mul_assoc _ _ (n ^ (q - m + 1):ℚ), ←one_div, mul_one_div,
+      div_div_eq_div_mul, hqm1],
+    rw [nat.cast_dvd],
+    { rw [nat.cast_mul],
+      ring, },
+    { simp only [nat.factorial_mul_factorial_dvd_factorial (le_of_lt h)], },
+    { simp only [ne.def, nat.cast_eq_zero, mul_eq_zero, nat.factorial_ne_zero, not_false_iff,
+        or_self], },
+  },
+end
+
+lemma aux_power_series_equal (n:ℕ):
+  (exp ℚ - 1) * power_series.mk (λ p, (finset.range n).sum (λ k,
+  (k:ℚ)^p * algebra_map ℚ ℚ p.factorial⁻¹)) =
+  (exp ℚ - 1) * power_series.mk (λ p, (finset.range (p + 1)).sum(λ i,
+  (bernoulli i)*((p + 1).choose i)*n^(p + 1 - i)/((p + 1).factorial))) :=
+begin
+  rw [←exp_pow_sum n],
+  have h_geom_sum: (geom_series (exp ℚ) n) * (exp ℚ - 1) = (exp ℚ)^n - 1 := geom_sum_mul _ _,
+  simp [geom_series] at h_geom_sum,
+  rw [mul_comm, h_geom_sum],
+  have h_const:  C ℚ (constant_coeff ℚ ((exp ℚ)^n))  = 1,
+    { simp only [one_pow, constant_coeff_exp, ring_hom.map_pow, ring_hom.map_one],},
+  have h_r: (exp ℚ)^n - 1 = X * mk (λ (p : ℕ), (coeff ℚ (p + 1)) (exp ℚ ^ n)),
+    { rw [←h_const], rw [sub_const_eq_X_mul_shift],},
+  rw [h_r, ←bernoulli_power_series, mul_assoc, aux_cauchy_prod],
+end
+
+theorem faulhaber (n p:ℕ):
+(finset.range n).sum(λk, (k:ℚ)^p) =
+(finset.range (p + 1)).sum(λ i,
+ (-1)^i*(bernoulli' i)*((p + 1).choose i)*n^(p + 1 - i)/(p + 1)) :=
+begin
+  have hpseq: power_series.mk (λ p, (finset.range n).sum (λ k,
+    (k:ℚ)^p * algebra_map ℚ ℚ p.factorial⁻¹)) =
+    power_series.mk (λ p, (finset.range (p + 1)).sum(λ i,
+      (-1)^i*(bernoulli' i)*((p + 1).choose i)*n^(p + 1 - i)/((p + 1).factorial))),
+  { have hexp: exp ℚ - 1 ≠ 0,
+    { rw [exp],
+      simp only [power_series.ext_iff, linear_map.map_zero, one_div, coeff_mk, coeff_one,
+      ring_hom.id_apply, linear_map.map_sub, ne.def, not_forall,
+      rat.algebra_map_rat_rat],
+      use 1,
+      simp only [nat.factorial_one, sub_zero, if_false, inv_one, not_false_iff,
+      one_ne_zero, nat.cast_one], },
+    apply mul_left_cancel' hexp (aux_power_series_equal n), },
+  simp only [power_series.ext_iff, cast_succ, coeff_mk, cast_mul, id_apply, rat.algebra_map_rat_rat]
+    at hpseq,
+  have h: (finset.range n).sum (λ k, (k:ℚ)^p * p.factorial⁻¹) = (finset.range (p + 1)).sum(λ i,
+  (-1)^i*(bernoulli' i)*((p + 1).choose i)*n^(p + 1 - i)/((p + 1).factorial)) := hpseq p,
```

Comment:
There must be a better way of saying " replace `hpseq `, with `hpseq p`, without repeating everyting, also above.

Reply:
It seems like you could use `have h := hpseq p` , or `specialize hpseq with p` , (the latter works if you won't be using `hpseq` again).

Reply:
`specialize` was exactly what I was looking for, thanks!